File: app/src/main/java/fr/geobert/radis/tools/ToggleImageButton.java
/**
 * From https://gist.github.com/akshaydashrath/9662072
 */

package fr.geobert.radis.tools;

import android.content.Context;
import android.content.res.TypedArray;
import android.os.Bundle;
import android.os.Parcelable;
import android.util.AttributeSet;
import android.widget.Checkable;
import android.widget.ImageButton;
import fr.geobert.radis.R;

public class ToggleImageButton extends ImageButton implements Checkable {

    private static final int[] CHECKED_STATE_SET = {
            android.R.attr.state_checked
    };
    private boolean isChecked;
    private boolean isBroadCasting;
    private OnCheckedChangeListener onCheckedChangeListener;

    public ToggleImageButton(Context context) {
        super(context);
    }

    public ToggleImageButton(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public ToggleImageButton(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        initAttr(context, attrs, defStyle);
    }

    private void initAttr(Context context, AttributeSet attrs, int defStyle) {
        TypedArray a =
                context.obtainStyledAttributes(
                        attrs, R.styleable.ToogleImageButton, defStyle, 0);
        boolean checked = a.getBoolean(R.styleable.ToogleImageButton_checked, false);
        setChecked(checked);
        a.recycle();
    }

    @Override
    protected void onFinishInflate() {
        super.onFinishInflate();
    }

    @Override
    public boolean performClick() {
        toggle();
        return super.performClick();
    }

    @Override
    public boolean isChecked() {
        return isChecked;  //To change body of implemented methods use File | Settings | File Templates.
    }

    @Override
    public void setChecked(boolean checked) {
        if (isChecked == checked) {
            return;
        }
        isChecked = checked;
        refreshDrawableState();
        if (isBroadCasting) {
            return;
        }
        isBroadCasting = true;
        if (onCheckedChangeListener != null) {
            onCheckedChangeListener.onCheckedChanged(this, isChecked);
        }
        isBroadCasting = false;
    }

    @Override
    public int[] onCreateDrawableState(int extraSpace) {
        final int[] drawableState = super.onCreateDrawableState(extraSpace + 1);
        if (isChecked()) {
            mergeDrawableStates(drawableState, CHECKED_STATE_SET);
        }
        return drawableState;
    }

    public void setOnCheckedChangeListener(OnCheckedChangeListener onCheckedChangeListener) {
        this.onCheckedChangeListener = onCheckedChangeListener;
    }

    @Override
    public void toggle() {
        setChecked(!isChecked);
    }

    @Override
    public Parcelable onSaveInstanceState() {
        Bundle bundle = new Bundle();
        bundle.putParcelable("instanceState", super.onSaveInstanceState());
        bundle.putBoolean("state", isChecked);
        return bundle;
    }

    @Override
    protected void onRestoreInstanceState(Parcelable state) {
        if (state instanceof Bundle) {
            Bundle outState = (Bundle) state;
            setChecked(outState.getBoolean("state"));
            state = outState.getParcelable("instanceState");
        }
        super.onRestoreInstanceState(state);
    }

    public interface OnCheckedChangeListener {

        void onCheckedChanged(ToggleImageButton buttonView, boolean isChecked);
    }
}

File: app/src/main/kotlin/fr/geobert/radis/tools/ToggleImageButton.kt
/**
 * From https://gist.github.com/akshaydashrath/9662072
 */

package fr.geobert.radis.tools

import android.content.Context
import android.os.Bundle
import android.os.Parcelable
import android.util.AttributeSet
import android.view.View
import android.widget.Checkable
import android.widget.ImageButton
import fr.geobert.radis.R

class ToggleImageButton : ImageButton, Checkable {
    private var isChecked: Boolean = false
    private var isBroadCasting: Boolean = false
    private var onCheckedChangeListener: OnCheckedChangeListener? = null

    constructor(context: Context) : super(context) {
    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
    }

    constructor(context: Context, attrs: AttributeSet, defStyle: Int) : super(context, attrs, defStyle) {
        initAttr(context, attrs, defStyle)
    }

    private fun initAttr(context: Context, attrs: AttributeSet, defStyle: Int) {
        val a = context.obtainStyledAttributes(
                attrs, R.styleable.ToogleImageButton, defStyle, 0)
        val checked = a.getBoolean(R.styleable.ToogleImageButton_checked, false)
        setChecked(checked)
        a.recycle()
    }

    override fun onFinishInflate() {
        super.onFinishInflate()
    }

    override fun performClick(): Boolean {
        toggle()
        return super.performClick()
    }

    override fun isChecked(): Boolean {
        return isChecked  //To change body of implemented methods use File | Settings | File Templates.
    }

    override fun setChecked(checked: Boolean) {
        if (isChecked == checked) {
            return
        }
        isChecked = checked
        refreshDrawableState()
        if (isBroadCasting) {
            return
        }
        isBroadCasting = true
        if (onCheckedChangeListener != null) {
            onCheckedChangeListener!!.onCheckedChanged(this, isChecked)
        }
        isBroadCasting = false
    }

    override fun onCreateDrawableState(extraSpace: Int): IntArray {
        val drawableState = super.onCreateDrawableState(extraSpace + 1)
        if (isChecked()) {
            View.mergeDrawableStates(drawableState, CHECKED_STATE_SET)
        }
        return drawableState
    }

    fun setOnCheckedChangeListener(onCheckedChangeListener: OnCheckedChangeListener) {
        this.onCheckedChangeListener = onCheckedChangeListener
    }

    override fun toggle() {
        setChecked(!isChecked)
    }

    public override fun onSaveInstanceState(): Parcelable {
        val bundle = Bundle()
        bundle.putParcelable("instanceState", super.onSaveInstanceState())
        bundle.putBoolean("state", isChecked)
        return bundle
    }

    override fun onRestoreInstanceState(s: Parcelable) {
        var state = s
        if (state is Bundle) {
            setChecked(state.getBoolean("state"))
            state = state.getParcelable<Parcelable>("instanceState")
        }
        super.onRestoreInstanceState(state)
    }

    interface OnCheckedChangeListener {

        fun onCheckedChanged(buttonView: ToggleImageButton, isChecked: Boolean)
    }

    companion object {

        private val CHECKED_STATE_SET = intArrayOf(android.R.attr.state_checked)
    }
}
